Add search condition for getting articles

JosephT5566 opened the pull request:
Hi
我試著增加了一些features，
再麻煩你有空的時候幫忙review一下看看是否合適~~
Thanks

## Changes
1. 在getArticles()中新增search條件(by push, author or title)
2. 新增相對應的unit test
3. 新增launch.json來實現mocha debug

## Note
1. 目前中文的關鍵字title搜尋無法成功，不清楚是不是編碼的問題，還需要trace一下
2. mocha debug的部分用起來感覺還是不太穩定，不知道是不是遺漏了什麼QQ

JosephT5566 committed:
[Merge]
1. Add debug option for mocha
2. Add search feature and correspond unit test.

kevinptt0323 commented on src/sites/ptt/bot.js:
```diff
@@ -20,6 +20,11 @@ class Bot extends EventEmitter {
     'message',
     'error',
   ];
+  searchCondition = {
```
改成在 `constructor` 裡面呼叫 `this.resetSearchCondition`，可以維持邏輯一致性。

JosephT5566 replied:
我後來發現可以把`searchCondition`搬到在`constructor`中建立，
不過你的意思是還要在`constructor`中再執行一次`this.resetSearchCondition`嗎

kevinptt0323 replied:
目前這樣宣告跟放到 `constructor` 意思差不多，影響不大。
我的意思是，假如今天 `searchCondition` 內的結構改變的話，初始化跟 `resetSearchCondition` 都要更新，不如利用 `resetSearchCondition` 來做初始化，應該會比較單純。

kevinptt0323 replied:
比如未來 `searchCondition` 有 `key1`、`key2`、…、`keyn`，那初始化跟 `resetSearchCondition` 就都要寫 n 行相同的邏輯。除非有什麼是初始化和 reset 行為可能會不同的 case，你也可以提出來～

JosephT5566 replied:
原來如此~~謝謝解釋!!
我合併了初始化方法，並改成用初始化來完成`resetSearchCondition`。

kevinptt0323 commented on src/sites/ptt/bot.js:
```diff
@@ -20,6 +20,11 @@ class Bot extends EventEmitter {
     'error',
   ];
+  searchCondition = {
+    searchType: "",
```
叫做 `type` 就可以了?

JosephT5566 replied:
好啊

kevinptt0323 replied:
小寫 `type`

kevinptt0323 commented on src/sites/ptt/bot.js:
```diff
@@ -188,8 +193,19 @@ class Bot extends EventEmitter {
     return authorArea === "作者";
   }
 
+  searchConditionIsSet() {
```
`isSearchConditionSet`

JosephT5566 replied:
OK

kevinptt0323 commented on src/sites/ptt/bot.js:
```diff
@@ -398,4 +414,26 @@ class Bot extends EventEmitter {
   }
 }
 
+Bot.prototype.setSearchCondition = function (searchType, condition) {
```
是否可以把 prototype function 移到 class 裡面宣告？

JosephT5566 replied:
OK

kevinptt0323 commented on test/articles.js:
```diff
@@ -18,6 +18,138 @@ const newbot = async () => {
+    });
+    let articles;
+    it('should get correct articles with specified author name from board', async () => {
+      ptt.setSearchCondition('author', 'Gaiaesque');
```
將 `'Gaiaesque'` 抽換成一個 variable 先宣告，這樣下面就用同一個 variable，也比較好修改測試時使用的參數。

JosephT5566 replied:
OK~

kevinptt0323 reviewed (commented):
首先非常感謝你的 PR 😃 

Mocha Test 的部分我也常常跑到壞掉，不知道是不是跟 PTT 連線機制有關係。

這個 PR 看起來不支援多重搜尋（例如 `/JOJO /心得 Z20` 之類的），也許我們之後可以再新增。

vscode 我自己是沒有在用，請問可以改成呼叫 `npm run test` 嗎？timeout 等參數你可以修改 `package.json`。

JosephT5566 commented:
不確定欸，不知道是不是因為太頻繁登入之類的關係，有時候我卡住我就先把確定pass的給`skip`掉就能繼續執行了。
對欸！我沒想到多重搜尋這一層，或許後許再來新增~
好~謝謝你的建議，後續再試試看一些參數設定，不過debug功能不太靈光，我現在也只是拿他來跑單一測試檔(或許`launch.json`就不要更新上來?)。

JosephT5566 committed:
[Modify] Modify issues

kevinptt0323 commented on src/sites/ptt/bot.js:
```diff
@@ -193,17 +193,39 @@ class Bot extends EventEmitter {
+  }
+  
+  resetSearchCondition() {
+    this.searchCondition.Type = "";
```
小寫 `type`

JosephT5566 replied:
我疏忽了~

kevinptt0323 reviewed (changes requested):
偶爾突然測不過我暫時也沒解法。
`launch.json` 看你要不要這次先拿掉，之後有更好的 config 可以再發 PR 之類的。
單一搜尋完成我就 approve 了。

JosephT5566 committed:
[Modify] Modify issues and codes
1. execute resetSearchCondition() by searchCondition.init()
2. change searchCondition.Type to searchCondition.type
3. remove launch.json
4. modify isSearchConditionSet() method

JosephT5566 commented:
先把`launch.json`拿掉好了，之後有更好的作法的話再提出~
